In [155]:
from IPython.display import HTML

HTML('''<script>
code_show=true;
function code_toggle() {
 if (code_show){
 $('div.input').hide();
 } else {
 $('div.input').show();
 }
 code_show = !code_show
}
$( document ).ready(code_toggle);
</script>
<form action="javascript:code_toggle()"><input type="submit" value="Click here to toggle on/off the raw code."></form>''')

--- 
# Model Fitting and Techniques 
***

The overall goal of this section is to try various techniques to fit a model for mortality rate using food consumption data.  First, we will find a null model, representing the 'average' input and representing a baseline estimation that we will then improve upon. Then we will fit a multilinear regression to all of the predictors (all livestock and all crop predictors), and find a cross-validated $R^2$ for this naive model. Next, we will try more advanced techniques such as Lasso, Ridge, Step-wise, and Regression Trees to improve this model. 

To summarize, our null model achieved a cross-validated $R^2$ score of 0 for all three diseases. Our naive model achieved a cross-validated score of $ $ for diabetes, $ $ for cancer, and $ $ for cardiovascular diseases.

## Null Model

Before fitting the linear regression, we will find a simple null model for global food consumption data. To calculate the null model, we found the average of each predictor column in the Dataframe. This gives us a 'global average' of consumption of each predictor. We can then use the null model to establish a baseline $R^2$ that we will then improve upon using our linear regression models.

In [101]:
import numpy as np
import matplotlib 
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import matplotlib.cm as cmx
import matplotlib.colors as colors
import pandas as pd
import math
from sklearn.cross_validation import cross_val_score, LeaveOneOut
from sklearn.cross_validation import KFold
from sklearn.linear_model import LinearRegression as LinReg
from sklearn.cross_validation import train_test_split as sk_split
import statsmodels.api as sm

%matplotlib inline

In [2]:
# Read in initial dataframe
x_df = pd.read_csv('predictors_filled_156.csv')

# read in disease rates
diabetes_df = pd.read_csv('diabetes_156.csv',index_col = 0)
cardio_df = pd.read_csv('cardio_156.csv',index_col = 0)
cancer_df= pd.read_csv('cancer_156.csv',index_col = 0)

### Null Model testing:

As expected, testing the null model on various training set give us a cross-validated $R^2$ of approximately zero for all three diseases. 

#### Cancer: 
The null model for cancer will always predict the mean cancer mortality rate. Testing on cancer, we get an $R^2$ of 3.33 E -16, which is ~ 0:

In [3]:
# Null Model Cancer
null_model = LinReg()
null_model.fit(x_df, [np.mean(cancer_df['Cancer Mortality Rate'])]*x_df.shape[0])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [4]:
# Test Cancer.
null_model.score(x_df, cancer_df)

0.0

#### Diabetes
Testing on diabetes, we also get an $R^2$ of 0.

In [5]:
# Fit Diabetes Null Model
null_model = LinReg()
null_model.fit(x_df, [np.mean(diabetes_df['Diabetes Mortality Rate'])]*x_df.shape[0])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [6]:
# Test Diabetes.
null_model.score(x_df, diabetes_df)

0.0

#### Test Cardiovascular Diseases

In [7]:
# Test cardiovascular diseases
null_model = LinReg()
null_model.fit(x_df, [np.mean(cardio_df['Cardio Mortality Rate'])]*x_df.shape[0])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [8]:
# Test cardiovascular diseases.
null_model.score(x_df, cardio_df)

0.0

# Simple LinReg

## Cancer LinReg

Now, we will fit a simple multi-linear regression to all of the food consumption inputs for each of the diseases. First, for cancer, our regression has an initial $R^2$ on the training set of .85, and a cross-validated $R^2$ of -16.5 for $k = 5$ folds. 

In [9]:
linreg = LinReg()
linreg.fit(x_df, cancer_df)
print "Training r^2:",linreg.score(x_df, cancer_df)

Training r^2: 0.842393103733


In [10]:
# Cross validated R-squared score
np.mean(cross_val_score(LinReg(), x_df,cancer_df, cv = KFold(151, 5), scoring = "r2"))

-30.410282842295079

To further examine the accuracy of this model, the map below displays the fractional difference of the model estimates as compared to the actual cancer data on a world map. As we can see, the vast majority of countries are colored a dark blue/ purple color, indicating they have a low fractional difference. Countries colored a brighter purple/pink color indicate an overestimate, while countries colored in a brighter blues indicate an underestimate.

In [11]:
# PUT GRAPH HERE

## Diabetes LinReg
For diabetes, our regression has an initial cross-validated $R^2$ of ** PUT THE R^2 HERE**. 

In [79]:
linreg = LinReg()
linreg.fit(x_df, diabetes_df)
print "Training R^2", linreg.score(x_df, diabetes_df)

# Cross-validated R^2 score for diabetes
print "CV R^2 score:",np.mean(cross_val_score(LinReg(), x_df,diabetes_df, cv = KFold(15, shuffle = True), scoring = "r2"))

Training R^2 0.807019987902
CV R^2 score: -1.88860041785


Again, we can examine a world map to see the fractional differences. Again, dark blue/purple colors indicate an accurate estimate,brighter purples/pinks indicate an overestimate, and brighter blues indicate an underestimate. We note thatthis model has slightly worse performance than our cancer model, which may have to do with the fact that diabetes does not lead to death as commonly as cancer does.

## Cardiovascular Diseases LinReg
For diabetes, our regression has an initial $R^2$ on the training set of .856, and a cross-validated $R^2$ of -6.02

In [13]:
linreg = LinReg()
linreg.fit(x_df, cardio_df)
print "Training R^2", linreg.score(x_df, cardio_df)
print "Cross-validated R^2", np.mean(cross_val_score(LinReg(), x_df,cardio_df, cv = KFold(151, 5), scoring = "r2"))

Training R^2 0.850938184461
Cross-validated R^2 -5.2779763269


Again, we can examine a world map to see the fractional differences. Dark blue/purple colors indicate an accurate estimate,brighter purples/pinks indicate an overestimate, and brighter blues indicate an underestimate. While this model appears to be fairly accurate for *THESE COUNTRIES*, it could be improved for *THESE*. This might be due to certain predictors, such as *SOME RANDOM PREDICTOR*, that is more heavily weighted for larger countries than for the country that is seeing a larger fractional difference.

--- 
# Advanced Models 
***

In this section, we will use various other regression techniques and variable selection techniques to attempt to improve upon our naive model. In particular, we will try 

1. Lasso
2. PCA
3. Regression Tree
4. Step-wise Variable Selection

For reference, our naive model gives us the following cross-validated $R^2$ values with $k = 5$: 

||Cardio | Diabetes | Cancer
|--- | --- | --- | ---|
|R^2 (Training)| .856 | .834 | .856|

## Lasso

The naive model brought up in the previous section has one major flaw: by including all of the predictors, it is very likely to be overfitted to the initial dataset. As such, we would like to reduce that overfitting by using variable selection techniques such as Lasso to reduce the number of predictors our model includes. 
Using the LassoCV package in sklearn, we obtain the following cross-validated $R^2$:

| |Cardio   |  Diabetes | Cancer  |
|-----|---|---|---|
|$r^2$ (Lasso) |  .456 |  .31 |   .053|


Below is a function we used to calculate the cross-validated r^2 for lasso over a number of folds for a certain parameter value.

In [14]:
def lasso_k_fold_r_squared(x_train, y_train, num_folds, param_val):
    n_train = x_train.shape[0]
    n = int(np.round(n_train * 1. / num_folds)) # points per fold

    # Iterate over folds
    cv_r_squared = 0
    
    for fold in range(1, num_folds + 1):
        # Take k-1 folds for training 
        x_first_half = x_train.iloc[:n * (fold - 1), :]
        x_second_half = x_train.iloc[n * fold + 1:, :]
        x_train_cv = np.concatenate((x_first_half, x_second_half), axis=0)
        
        y_first_half = y_train.iloc[:n * (fold - 1)]
        y_second_half = y_train.iloc[n * fold + 1:]
        y_train_cv = np.concatenate((y_first_half, y_second_half), axis=0)
        
        # Take the middle fold for testing
        x_test_cv = x_train.iloc[1 + n * (fold - 1):n * fold, :]
        y_test_cv = y_train.iloc[1 + n * (fold - 1):n * fold]

        # Fit Decision Tree model with parameter value on CV train set, and evaluate CV test performance
        reg = Lasso(alpha = param_val, normalize=True)
        reg.fit(x_train_cv, y_train_cv)
        coefficients = reg.coef_
        #print len([i for i, item in enumerate(coefficients) if abs(item) >0])
        r_squared = reg.score(x_test_cv, y_test_cv)
    
        # Cummulative R^2 value across folds
        cv_r_squared += r_squared

    # Return average R^2 value across folds
    return cv_r_squared * 1.0 / num_folds

### Cancer

Running Lasso on different values of alpha yields a top cross-validated $r^2$ score of .059, for alpha = .1

In [15]:
from sklearn.linear_model import Lasso
from sklearn.linear_model import LassoCV

print "LASSO"
for alpha in [.001, .01, .1,.4,.5,.6,1,5,10,100]:
    print "cancer",alpha, lasso_k_fold_r_squared(x_df,cancer_df,5, alpha)

LASSO
cancer 0.001 -10.0863952587
cancer 0.01

/Users/Susan/anaconda/lib/python2.7/site-packages/sklearn/linear_model/coordinate_descent.py:466: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations
  ConvergenceWarning)


 -2.62319418962
cancer 0.1 -0.189052791559
cancer 0.4 0.15663394152
cancer 0.5 0.13439828752
cancer 0.6 0.117227909906
cancer 1 0.0530720050382
cancer 5 -0.0745935162681
cancer 10 -0.0745935162681
cancer 100 -0.0745935162681


In [ ]:
# LOOCV 

def loocv_score(x_df,y_df):
    LeaveOneOut
    

### Cardiovascular Diseases
Running Lasso on different values of alpha yields a top cross-validated $r^2$ score of .459, for alpha = .9

In [158]:
print "LASSO"
for folds in range(5, 32):
    if folds == 27 or folds == 28:
        continue
    print folds,"-FOLD \n"
    for alph in [.1, .5,.8, .9, 1, 2,5, 10]:
        print "cardio",alph, lasso_k_fold_r_squared(x_df,cardio_df,folds, alph)

 LASSO
5 -FOLD 

cardio 0.1 -0.111196584756
cardio 0.5 0.415841480013
cardio 0.8 0.451848911048
cardio 0.9 0.455639379888
cardio 1 0.456181146633
cardio 2 0.404302118408
cardio 5 0.0132238574549
cardio 10 -0.010213535932
6 -FOLD 

cardio 0.1 0.0725073431946
cardio 0.5 0.402144673879
cardio 0.8 0.45531395782
cardio 0.9 0.457284742801
cardio 1 0.455753776186
cardio 2 0.405437627519
cardio 5 -0.00170633655146
cardio 10 -0.0152504500031
7 -FOLD 

cardio 0.1 0.0694475897836
cardio 0.5 0.449822469823
cardio 0.8 0.465917040409
cardio 0.9 0.466196311375
cardio 1 0.465317817456
cardio 2 0.401359065679
cardio 5 -0.0288567060725
cardio 10 -0.0389712469074
8 -FOLD 

cardio 0.1 -0.0913623843405
cardio 0.5 0.424231562168
cardio 0.8 0.454902852008
cardio 0.9 0.458586547056
cardio 1 0.459338572343
cardio 2 0.388306847732
cardio 5 -0.0402185999135
cardio 10 -0.0459111114413
9 -FOLD 

cardio 0.1 -0.111700183753
cardio 0.5 0.344850695317
cardio 0.8 0.411425851844
cardio 0.9 0.415938676883
cardio 1 0.4151

In [159]:
print np.mean(cross_val_score(Lasso(alpha = 1), x_df, cardifo_df, cv = LeaveOneOut(156), scoring = 'mean_squared_error'))

-15291.791902


In [17]:
reg = Lasso(alpha = .9, normalize=True)
reg.fit(x_df, cardio_df)

Lasso(alpha=1, copy_X=True, fit_intercept=True, max_iter=1000, normalize=True,
   positive=False, precompute=False, random_state=None, selection='cyclic',
   tol=0.0001, warm_start=False)

In [151]:
from sklearn.metrics import r2_score
def loocv_score(x_df, y_df, reg,n):
    mse_tot = 0
    r2_tot = 0
    for train, test in LeaveOneOut(n):
        # Get train, test indices from LOO
        x_train = x_df.iloc[train, :]
        y_train = y_df.iloc[train, :]
        x_test = x_df.iloc[test, :]
        y_test = y_df.iloc[test, :]
        
        # Fit the regression to train
        reg.fit(x_train, y_train)
        r2_tot += r2_score(y_test, reg.predict(x_test))
        y_test - reg.predict(x_test)
        mse_tot += ((y_test - reg.predict(x_test)).iloc[0, 0])**2
    return mse_tot*1.0/n

In [153]:
for alph in [.001, .01, .05, .1, .5, 1, 2, 10, 100]: 
    print loocv_score(x_df, cardio_df, Lasso(alpha = alph), 156)

42698.1526592
40518.5416752
34399.0222303
30439.5567253
19315.9852061
15291.791902
13066.2740138
11733.3050556
8563.52034934


### Diabetes 
Running Lasso on different values of alpha from .001 to 100 yields a top cross-validated $r^2$ score of .29, for alpha = .1

In [18]:
print "LASSO"
for alpha in [.001, .01, .05, .1, .5, 1, 10, 100]:
    print "diabetes",alpha, lasso_k_fold_r_squared(x_df,diabetes_df,4, alpha)

LASSO
diabetes 0.001 -8.75554668254
diabetes 0.01 -1.98857707569
diabetes 0.05 0.172208052135
diabetes 0.1 0.317952766042
diabetes 0.5 0.228773331487
diabetes 1 0.0493154877338
diabetes 10 -0.0130465434519
diabetes 100 -0.0130465434519


## Ridge 

Again, in this section we'd like to try to use Ridge regression to improve the cross-validated $r^2$ for our model and reduce overfitting of the model.
Below is a function we used to calculate the cross-validated r^2 for ridge regression over a number of folds for a certain parameter value.

To summarize, we have:


| |Cardio   |  Diabetes | Cancer  |
|-----|---|---|---|
|$r^2$ (Ridge) |  .424 |  .323 |   .168|

In [19]:
from sklearn.linear_model import Ridge
def ridge_k_fold_r_squared(x_train, y_train, num_folds, param_val):
    n_train = x_train.shape[0]
    n = int(np.round(n_train * 1. / num_folds)) # points per fold

    # Iterate over folds
    cv_r_squared = 0
    
    for fold in range(1, num_folds + 1):
        # Take k-1 folds for training 
        x_first_half = x_train.iloc[:n * (fold - 1), :]
        x_second_half = x_train.iloc[n * fold + 1:, :]
        x_train_cv = np.concatenate((x_first_half, x_second_half), axis=0)
        
        y_first_half = y_train.iloc[:n * (fold - 1)]
        y_second_half = y_train.iloc[n * fold + 1:]
        y_train_cv = np.concatenate((y_first_half, y_second_half), axis=0)
        
        # Take the middle fold for testing
        x_test_cv = x_train.iloc[1 + n * (fold - 1):n * fold, :]
        y_test_cv = y_train.iloc[1 + n * (fold - 1):n * fold]

        # Fit Decision Tree model with parameter value on CV train set, and evaluate CV test performance
        reg = Ridge(alpha = param_val, normalize=True)
        reg.fit(x_train_cv, y_train_cv)
        r_squared = reg.score(x_test_cv, y_test_cv)
    
        # Cummulative R^2 value across folds
        cv_r_squared += r_squared

    # Return average R^2 value across folds
    return cv_r_squared * 1.0 / num_folds

### Cardiovascular Diseases
Running Ridge regression on different values of alpha from .001 to 100 yields a top cross-validated $r^2$ score of .425, for alpha = 1.1

In [20]:
print "RIDGE"
for alpha in [.001, .01, .1, .9,1,1.1,1.3,1.5,1.9,5, 10, 11, 100]:
    print "cardio",alpha, ridge_k_fold_r_squared(x_df,cardio_df,5, alpha)

RIDGE
cardio 0.001 -4.9919143728
cardio 0.01 -2.75443505612
cardio 0.1 -0.175066648391
cardio 0.9 0.404437744071
cardio 1 0.410643001
cardio 1.1 0.415244083252
cardio 1.3 0.421038145156
cardio 1.5 0.423729737401
cardio 1.9 0.423597298657
cardio 5 0.36475407367
cardio 10 0.27837000173
cardio 11 0.265254747482
cardio 100 0.0444563936422


### Diabetes

Running Lasso on different values of alpha from .001 to 100 yields a top cross-validated $r^2$ score of .317, for alpha = 1

In [21]:
print "RIDGE"
for alpha in [.001, .01, .1, .9,1,3, 5, 10,100, 1000]:
    print "diabetes",alpha, ridge_k_fold_r_squared(x_df,diabetes_df,4, alpha)

RIDGE
diabetes 0.001 -9.77735177572
diabetes 0.01 -3.76622924788
diabetes 0.1 -0.279553857695
diabetes 0.9 0.314979301771
diabetes 1 0.321214840769
diabetes 3 0.325946173079
diabetes 5 0.300072240351
diabetes 10 0.246363686822
diabetes 100 0.0540094335668
diabetes 1000 -0.00498501391727


In [22]:
reg = Ridge(alpha = 1, normalize=True)
reg.fit(x_df, diabetes_df)

Ridge(alpha=1, copy_X=True, fit_intercept=True, max_iter=None, normalize=True,
   random_state=None, solver='auto', tol=0.001)

### Cancer

Running Lasso on different values of alpha from .001 to 100 yields a top cross-validated $r^2$ score of .18, for alpha = 5

In [23]:
print "RIDGE"

for alpha in [.001, .01, .1, 1, 3, 4,5,10, 20, 1000]:
    print "cancer",alpha, ridge_k_fold_r_squared(x_df,cancer_df,5, alpha)

RIDGE
cancer 0.001 -9.11931708715
cancer 0.01 -3.21130503559
cancer 0.1 -0.681543066623
cancer 1 0.0574910468362
cancer 3 0.162757999505
cancer 4 0.168501040516
cancer 5 0.168009123824
cancer 10 0.144581726236
cancer 20 0.0992315721652
cancer 1000 -0.0668769303926


---
## Regression Trees
---

In this section, we'll try regression trees to see if they improve our model.

Below, we have a function that we'll use to find the $r^2$ value for a given number of folds and certain hyperparameter.

In [24]:
from sklearn.tree import DecisionTreeRegressor
def rtree_k_fold_r_squared(x_train, y_train, num_folds, param_val):
    n_train = x_train.shape[0]
    n = int(np.round(n_train * 1. / num_folds)) # points per fold

    # Iterate over folds
    cv_r_squared = 0
    
    for fold in range(1, num_folds + 1):
        # Take k-1 folds for training 
        x_first_half = x_train.iloc[:n * (fold - 1), :]
        x_second_half = x_train.iloc[n * fold + 1:, :]
        x_train_cv = np.concatenate((x_first_half, x_second_half), axis=0)
        
        y_first_half = y_train.iloc[:n * (fold - 1)]
        y_second_half = y_train.iloc[n * fold + 1:]
        y_train_cv = np.concatenate((y_first_half, y_second_half), axis=0)
        
        # Take the middle fold for testing
        x_test_cv = x_train.iloc[1 + n * (fold - 1):n * fold, :]
        y_test_cv = y_train.iloc[1 + n * (fold - 1):n * fold]

        # Fit Decision Tree model with parameter value on CV train set, and evaluate CV test performance
        reg = DecisionTreeRegressor(max_depth=param_val)
        reg.fit(x_train_cv, y_train_cv)
        r_squared = reg.score(x_test_cv, y_test_cv)
    
        # Cummulative R^2 value across folds
        cv_r_squared += r_squared

    # Return average R^2 value across folds
    return cv_r_squared * 1.0 / num_folds

### Cardiovascular Diseases

For cardiovascular diseases, we see the best $r^2 = .44$ for max_depth = 2. 

In [25]:
for depth in [2, 3, 4, 5, 10, 50, 100]:
    print depth, rtree_k_fold_r_squared(x_df,cardio_df,5, depth)

2 0.424335403226
3 0.31037836342
4 0.138594642848
5 0.11320094408
10 0.101539041984
50 0.197917860836
100 0.212245738199


### Diabetes

For diabetes, we see the best $r^2$ of $.157$ for a max-depth of 2.

In [26]:
for depth in [2, 3, 4, 5, 10, 50, 100]:
    print depth, rtree_k_fold_r_squared(x_df,diabetes_df,5, depth)

2 0.135877261593
3 0.0668897439002
4 -0.0282723431846
5 -0.013424750927
10 -0.203544872678
50 -0.199689970513
100 -0.34393776059


### Cancer

All of the $r^2$ for decision trees on cancer were negative, indicating that these actually perform worse than the null model and as such are not useful models to examine.

In [27]:
for depth in [2, 3, 5, 8, 9, 10, 20, 50, 70, 100]:
    print depth, rtree_k_fold_r_squared(x_df,cancer_df,5, depth)

2 -0.908504830182
3 -1.04343525752
5 -1.0104804726
8 -1.13983776169
9 -1.18740341487
10 -1.01382875454
20 -1.17985787005
50 -0.962559418604
70 -1.25227865248
100 -1.32841685985
